testing.rst

```diff
@@ -11,377 +11,513 @@ using both functional and unit tests.
+    DATABASE_URL="mysql://USERNAME:PASSWORD@127.0.0.1:3306/DB_NAME?serverVersion=5.7"
+
+.. TODO Symfony CLI automatically appends `_test` somehow sometimes, right?
+
+This assumes that each developer/machine uses a different database for the
+tests. If the test set-up is the same on each machine, use the ``.env.test``
+file instead and commit it to the shared repository. Learn more about
+:ref:`using multiple .env files in Symfony applications <configuration-multiple-env-files>`.
+
+After that, you can create the test database and all tables using:
+
+.. code-block:: terminal
+
+    # create the test database
+    $ php bin/console --env=test doctrine:database:create
+
+    # create the tables/columns in the test database
+    $ php bin/console --env=test doctrine:schema:create
 
 .. tip::
 
-    The ``Crawler`` only works when the response is an XML or an HTML document.
-    To get the raw content response, call ``$client->getResponse()->getContent()``.
+    A common practice is to append the ``_test`` suffix to the original
+    database names in tests. If the database name in production is called
+    ``project_acme`` the name of the testing database could be
+    ``project_acme_test``.
+
+Resetting the Database Automatically Before each Test
+.....................................................
+
+Tests should be independent from each other to avoid side effects. For
+example, if some test modifies the database (by adding or removing an
+entity) it could change the results of other tests.
 
-The crawler integrates with the ``symfony/css-selector`` component to give you the
-power of CSS selectors to find content in a page. To install the CSS selector
-component, run:
+The `DAMADoctrineTestBundle`_ uses Doctrine transactions to let each test
+interact with an unmodified database. Install it using:
 
 .. code-block:: terminal
 
-    $ composer require --dev symfony/css-selector
+    $ composer require --dev dama/doctrine-test-bundle
 
-Now you can use CSS selectors with the crawler. To assert that the phrase
-"Hello World" is present in the page's main title, you can use this assertion::
+Now, enable it as a PHPUnit extension:
 
-    $this->assertSelectorTextContains('html h1.title', 'Hello World');
+.. code-block:: xml
 
-This assertion checks if the first element matching the CSS selector contains
-the given text. This assert calls ``$crawler->filter('html h1.title')``
-internally, which allows you to use CSS selectors to filter any HTML element in
-the page and check for its existence, attributes, text, etc.
+    <!-- phpunit.xml.dist -->
+    <phpunit>
+        <!-- ... -->
 
-The ``assertSelectorTextContains`` method is not a native PHPUnit assertion and is
-available thanks to the ``WebTestCase`` class.
+        <extensions>
+            <extension class="DAMA\DoctrineTestBundle\PHPUnit\PHPUnitExtension"/>
+        </extensions>
+    </phpunit>
 
-.. versionadded:: 4.3
+That's it! This bundle uses a clever trick: it begins a database
```

Comment:
It will probably break tests using Panther. Can this be disabled on a per-test basis?

Reply:
I don't believe so.

Reply:
> Can this be disabled on a per-test basis?

Theoretically its possible. But this would defeat the purpose a bit since then some tests will actually persist changes in the DB and potentially impact other tests.

In this case I think one should split the tests into different suites and only enable `dama/doctrine-test-bundle` for non-Panther tests maybe.